### ISSUE TITLE
plugin binds a UDP socket for every build and never releases it

### ISSUE THREAD

**Issue by lattwood** [DataDog/jenkins-datadog-plugin#75]
This behaviour will continue until Jenkins runs out of file descriptors, rendering it completely inoperable. Jenkins will then fill the disk at a very rapid rate with logs explaining it's out of file descriptors.

Here's the output of netstat and tcpdump confirming this.

## tcpdump
```
21:07:22.531862 IP (tos 0x0, ttl 64, id 58450, offset 0, flags [DF], proto UDP (17), length 90)
    127.0.0.1.39127 > 127.0.0.1.8125: [bad udp cksum 0xfe59 -> 0xb716!] UDP, length 62
        0x0000:  0000 0304 0006 0000 0000 0000 6f6e 0800  ............on..
        0x0010:  4500 005a e452 4000 4011 583e 7f00 0001  E..Z.R@.@.X>....
        0x0020:  7f00 0001 98d7 1fbd 0046 fe59 6a65 6e6b  .........F.Yjenk
        0x0030:  696e 732e 6a6f 622e 636f 6d70 6c65 7465  ins.job.complete
        0x0040:  643a 317c 637c 2362 7261 6e63 683a 6f72  d:1|c|#branch:or
        0x0050:  6967 696e 2f64 6576 656c 6f70 6d65 6e74  igin/development
        0x0060:  2c6a 6f62 3a55 492d 5141                 ,job:UI-QA
21:07:27.421957 IP (tos 0x0, ttl 64, id 59563, offset 0, flags [DF], proto UDP (17), length 110)
    127.0.0.1.34209 > 127.0.0.1.8125: [bad udp cksum 0xfe6d -> 0xfecf!] UDP, length 82
        0x0000:  0000 0304 0006 0000 0000 0000 0000 0800  ................
        0x0010:  4500 006e e8ab 4000 4011 53d1 7f00 0001  E..n..@.@.S.....
        0x0020:  7f00 0001 85a1 1fbd 005a fe6d 6a65 6e6b  .........Z.mjenk
        0x0030:  696e 732e 6a6f 622e 636f 6d70 6c65 7465  ins.job.complete
        0x0040:  643a 317c 637c 2362 7261 6e63 683a 7061  d:1|c|#branch:pa
        0x0050:  6769 6e61 7465 2d74 6865 2d6f 7267 616e  ginate-the-organ
        0x0060:  697a 6174 696f 6e73 2d6c 6973 742c 6a6f  izations-list,jo
        0x0070:  623a 496e 7472 616e 6574 2050 5273       b:Intranet.PRs
21:07:45.624593 IP (tos 0x0, ttl 64, id 62489, offset 0, flags [DF], proto UDP (17), length 113)
    127.0.0.1.33975 > 127.0.0.1.8125: [bad udp cksum 0xfe70 -> 0x6cfc!] UDP, length 85
        0x0000:  0000 0304 0006 0000 0000 0000 0000 0800  ................
        0x0010:  4500 0071 f419 4000 4011 4860 7f00 0001  E..q..@.@.H`....
        0x0020:  7f00 0001 84b7 1fbd 005d fe70 6a65 6e6b  .........].pjenk
        0x0030:  696e 732e 6a6f 622e 636f 6d70 6c65 7465  ins.job.complete
        0x0040:  643a 317c 637c 2362 7261 6e63 683a 6f72  d:1|c|#branch:or
        0x0050:  6967 696e 2f70 6167 696e 6174 652d 6f72  igin/paginate-or
        0x0060:  6761 6e69 7a61 7469 6f6e 732d 6c69 7374  ganizations-list
        0x0070:  2c6a 6f62 3a49 6e74 7261 6e65 7420 5052  ,job:Intranet.PR
        0x0080:  73                                       s
21:08:01.167220 IP (tos 0x0, ttl 64, id 64502, offset 0, flags [DF], proto UDP (17), length 117)
    127.0.0.1.35373 > 127.0.0.1.8125: [bad udp cksum 0xfe74 -> 0x8de8!] UDP, length 89
        0x0000:  0000 0304 0006 0000 0000 0000 0000 0800  ................
        0x0010:  4500 0075 fbf6 4000 4011 407f 7f00 0001  E..u..@.@.@.....
        0x0020:  7f00 0001 8a2d 1fbd 0061 fe74 6a65 6e6b  .....-...a.tjenk
        0x0030:  696e 732e 6a6f 622e 636f 6d70 6c65 7465  ins.job.complete
        0x0040:  643a 317c 637c 2362 7261 6e63 683a 6f72  d:1|c|#branch:or
        0x0050:  6967 696e 2f70 6167 696e 6174 652d 7468  igin/paginate-th
        0x0060:  652d 6f72 6761 6e69 7a61 7469 6f6e 732d  e-organizations-
        0x0070:  6c69 7374 2c6a 6f62 3a49 6e74 7261 6e65  list,job:Intrane
        0x0080:  7420 5052 73                             t.PRs
```

## netstat
The values in the Recv-Q column are due to me using netcat to prod the ports
```
root@ip-10-0-103-29:/var/lib/jenkins# netstat -nape | grep java | grep udp6
udp6       0      0 :::35373                :::*                                106        2770877827  20518/java
udp6    3840      0 :::60431                :::*                                106        2770521115  20518/java
udp6       0      0 :::5353                 :::*                                106        2770515125  20518/java
udp6       0      0 :::39127                :::*                                106        2770877557  20518/java
udp6       0      0 :::33848                :::*                                106        2770515122  20518/java
udp6     768      0 :::58444                :::*                                106        2770831899  20518/java
udp6       0      0 :::33975                :::*                                106        2770877724  20518/java
udp6       0      0 :::34209                :::*                                106        2770877594  20518/java
udp6    1536      0 :::34463                :::*                                106        2770624072  20518/java
udp6       0      0 :::34679                :::*                                106        2770559677  20518/java
unix  2      [ ]         STREAM     CONNECTED     2770016066 20518/java
unix  2      [ ]         STREAM     CONNECTED     2770506247 20518/java
```

As you can see, the source ports in tcpdump match up with bound ports in the netstat output. I don't expect Jenkins to require non-blocking statsd sends, perhaps that's the easiest way to fix this?

**Comment by ssbarnea** [comment]
Thanks for filing this, you saved me the embarrassment of taking down a big production server by installing this plugin.

**Comment by remh** [comment]
Thanks for the detailed feedback we'll investigate this ASAP.

**Comment by ssbarnea** [comment]
@remh thanks, it would really useful to know how dangerous is this bug. Until we get some more info I will not install the plugin to production instances.

**Comment by remh** [comment]
I guess it depends on how busy the plugin is used. For reference we use the plugin ourselves on a biggish jenkins server that has been running for 3 weeks. The jenkins process has around 14k open FDs. 
We'll work on a fix ASAP.

**Comment by remh** [comment]
I believe this will be fixed with this PR: https://github.com/DataDog/jenkins-datadog-plugin/pull/73

We'll do a release today or early next week latest.

**Comment by ssbarnea** [comment]
Thanks, I am currently running 0.5.5 on an empty instance with a job scheduled to run every 30 minutes, just to see if there is any leak. Once I have the new version I will be able to install it and test it also.

**Comment by JohnLZeller** [comment]
Fixed as of 0.5.6. Thanks for your patience, all!

https://github.com/jenkinsci/datadog-plugin/releases/tag/datadog-0.5.6
